File: ui/backend/schemas/pipeline.py
"""Pydantic schemas for Pipeline API.

Story 5-5: Lead Pipeline Status Tracking.
Provides request/response schemas for pipeline dashboard endpoints.

Schemas:
    - PipelineLeadSchema: Lead summary for pipeline view
    - PipelineSummarySchema: Status counts, conversion rate, followup count
    - PipelineLeadListResponse: Paginated lead list
    - LeadHistoryEntrySchema: Single activity record
    - LeadHistoryResponse: List of history entries
    - StatusUpdateRequest: Manual status transition request
    - StatusUpdateResponse: Status update confirmation
    - PipelineMetricsSchema: Conversion funnel, avg time, weekly trend
    - LeadDetailSchema: Full lead detail with relations
    - AddNoteRequest: Add manual note to lead
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class PipelineLeadSchema(BaseModel):
    """Lead summary for pipeline view.

    Compact representation of a lead for Kanban board display.
    """

    id: str = Field(..., description="Lead UUID")
    email: str = Field(..., description="Primary email address")
    first_name: str = Field(..., description="Contact first name")
    last_name: str = Field(..., description="Contact last name")
    company: str = Field(..., description="Company name")
    job_title: Optional[str] = Field(default=None, description="Role/position")
    status: str = Field(..., description="Pipeline status")
    source: str = Field(..., description="Lead source")
    score: float = Field(..., ge=0, le=100, description="Lead score (0-100)")
    last_contacted_at: Optional[datetime] = Field(
        default=None, description="Last outreach timestamp"
    )
    days_since_contact: Optional[int] = Field(
        default=None, description="Days since last contact"
    )
    needs_followup: bool = Field(
        default=False, description="Whether lead needs follow-up"
    )
    contact_count: int = Field(default=0, description="Number of outreach attempts")
    created_at: datetime = Field(..., description="Record creation time")

    model_config = {"from_attributes": True}


class PipelineSummarySchema(BaseModel):
    """Pipeline dashboard summary.

    Top-level metrics for the pipeline overview.
    """

    total_leads: int = Field(..., ge=0, description="Total leads in pipeline")
    status_counts: dict[str, int] = Field(
        ..., description="Lead count per pipeline status"
    )
    conversion_rate: float = Field(
        ..., ge=0, description="Conversion rate percentage"
    )
    avg_days_in_pipeline: float = Field(
        ..., ge=0, description="Average days leads spend in pipeline"
    )
    followup_count: int = Field(
        ..., ge=0, description="Number of leads needing follow-up"
    )

    model_config = {"from_attributes": True}


class PipelineLeadListResponse(BaseModel):
    """Paginated lead list response.

    Supports offset-based pagination with cursor.
    """

    leads: list[PipelineLeadSchema] = Field(
        ..., description="List of leads for current page"
    )
    total_count: int = Field(..., ge=0, description="Total matching leads")
    has_more: bool = Field(..., description="Whether more leads exist")
    next_cursor: Optional[str] = Field(
        default=None, description="Cursor for next page"
    )

    model_config = {"from_attributes": True}


class LeadHistoryEntrySchema(BaseModel):
    """Single lead activity record.

    Represents one entry in the lead's activity timeline.
    """

    id: str = Field(..., description="Activity UUID")
    lead_id: str = Field(..., description="Lead UUID")
    activity_type: str = Field(..., description="Activity type")
    description: str = Field(..., description="Human-readable description")
    timestamp: datetime = Field(..., description="When activity occurred")
    actor: Optional[str] = Field(default=None, description="Who performed the action")
    metadata: Optional[dict] = Field(
        default=None, description="Additional activity data"
    )

    model_config = {"from_attributes": True}


class LeadHistoryResponse(BaseModel):
    """Lead activity history response.

    Complete activity timeline for a single lead.
    """

    entries: list[LeadHistoryEntrySchema] = Field(
        ..., description="Activity entries ordered by timestamp desc"
    )

    model_config = {"from_attributes": True}


class StatusUpdateRequest(BaseModel):
    """Manual status transition request.

    Used by operators to manually advance or close leads.
    """

    new_status: str = Field(..., description="Target pipeline status")
    reason: Optional[str] = Field(
        default=None, description="Reason for status change"
    )
    note: Optional[str] = Field(
        default=None, description="Optional note about the transition"
    )


class StatusUpdateResponse(BaseModel):
    """Status update confirmation response."""

    success: bool = Field(..., description="Whether update succeeded")
    lead: PipelineLeadSchema = Field(..., description="Updated lead data")
    message: str = Field(..., description="Confirmation message")

    model_config = {"from_attributes": True}


class PipelineMetricsSchema(BaseModel):
    """Conversion funnel and trend metrics.

    Detailed pipeline performance data for metrics panel.
    """

    conversion_rate: float = Field(
        ..., ge=0, description="Conversion rate percentage"
    )
    total_leads: int = Field(..., ge=0, description="Total leads in pipeline")
    converted_count: int = Field(
        ..., ge=0, description="Number of converted leads"
    )
    avg_time_per_stage: dict[str, float] = Field(
        ..., description="Average days per pipeline stage"
    )
    weekly_trend: list[dict] = Field(
        ..., description="Weekly new vs converted counts"
    )

    model_config = {"from_attributes": True}


class LeadDetailSchema(PipelineLeadSchema):
    """Full lead detail with relations.

    Extends PipelineLeadSchema with enrichment data, outreach data,
    activity history, and email records.
    """

    linkedin_url: Optional[str] = Field(default=None, description="LinkedIn profile")
    website_url: Optional[str] = Field(default=None, description="Company website")
    phone: Optional[str] = Field(default=None, description="Phone number")
    country: Optional[str] = Field(default=None, description="Country/region")
    industry: Optional[str] = Field(default=None, description="Industry vertical")
    company_size: Optional[str] = Field(
        default=None, description="Employee count range"
    )
    enrichment_data: Optional[dict] = Field(
        default=None, description="Enrichment service data"
    )
    outreach_data: Optional[dict] = Field(
        default=None, description="Outreach template data"
    )
    tags: list[str] = Field(default_factory=list, description="Custom tags")
    notes: Optional[str] = Field(default=None, description="Free-form notes")
    assigned_to: Optional[str] = Field(
        default=None, description="Sales rep assignment"
    )
    next_followup_at: Optional[datetime] = Field(
        default=None, description="Scheduled follow-up time"
    )
    converted_at: Optional[datetime] = Field(
        default=None, description="Conversion timestamp"
    )
    lost_reason: Optional[str] = Field(default=None, description="Why lead was lost")
    updated_at: Optional[datetime] = Field(
        default=None, description="Last update time"
    )
    activities: list[LeadHistoryEntrySchema] = Field(
        default_factory=list, description="Activity history"
    )
    emails: list[dict] = Field(
        default_factory=list, description="Outreach email records"
    )

    model_config = {"from_attributes": True}


class AddNoteRequest(BaseModel):
    """Add manual note request."""

    note: str = Field(..., min_length=1, description="Note text")


__all__ = [
    "PipelineLeadSchema",
    "PipelineSummarySchema",
    "PipelineLeadListResponse",
    "LeadHistoryEntrySchema",
    "LeadHistoryResponse",
    "StatusUpdateRequest",
    "StatusUpdateResponse",
    "PipelineMetricsSchema",
    "LeadDetailSchema",
    "AddNoteRequest",
]

File: ui/frontend-react/src/types/pipeline.ts
/**
 * TypeScript types for Lead Pipeline Dashboard.
 *
 * Story 5-5: Mirrors backend Pydantic schemas for type-safe API communication.
 */

/**
 * Lead status enum matching backend LeadStatus.
 */
export enum LeadStatus {
  NEW = "new",
  RESEARCHING = "researching",
  QUALIFIED = "qualified",
  OUTREACH_PENDING = "outreach_pending",
  CONTACTED = "contacted",
  REPLIED = "replied",
  MEETING_SCHEDULED = "meeting_scheduled",
  NEGOTIATING = "negotiating",
  CONVERTED = "converted",
  LOST = "lost",
  NURTURE = "nurture",
}

/**
 * Status display configuration for UI.
 */
export interface StatusConfig {
  label: string;
  color: string;
  bgColor: string;
}

/**
 * Status display configurations.
 */
export const STATUS_CONFIGS: Record<LeadStatus, StatusConfig> = {
  [LeadStatus.NEW]: { label: "New", color: "text-blue-700", bgColor: "bg-blue-100" },
  [LeadStatus.RESEARCHING]: { label: "Researching", color: "text-indigo-700", bgColor: "bg-indigo-100" },
  [LeadStatus.QUALIFIED]: { label: "Qualified", color: "text-cyan-700", bgColor: "bg-cyan-100" },
  [LeadStatus.OUTREACH_PENDING]: { label: "Outreach Pending", color: "text-amber-700", bgColor: "bg-amber-100" },
  [LeadStatus.CONTACTED]: { label: "Contacted", color: "text-orange-700", bgColor: "bg-orange-100" },
  [LeadStatus.REPLIED]: { label: "Replied", color: "text-teal-700", bgColor: "bg-teal-100" },
  [LeadStatus.MEETING_SCHEDULED]: { label: "Meeting", color: "text-purple-700", bgColor: "bg-purple-100" },
  [LeadStatus.NEGOTIATING]: { label: "Negotiating", color: "text-pink-700", bgColor: "bg-pink-100" },
  [LeadStatus.CONVERTED]: { label: "Converted", color: "text-green-700", bgColor: "bg-green-100" },
  [LeadStatus.LOST]: { label: "Lost", color: "text-red-700", bgColor: "bg-red-100" },
  [LeadStatus.NURTURE]: { label: "Nurture", color: "text-yellow-700", bgColor: "bg-yellow-100" },
};

/**
 * Pipeline lead summary for list views.
 */
export interface PipelineLead {
  id: string;
  email: string;
  first_name: string;
  last_name: string;
  company: string;
  job_title: string | null;
  status: string;
  source: string;
  score: number;
  last_contacted_at: string | null;
  days_since_contact: number | null;
  needs_followup: boolean;
  contact_count: number;
  created_at: string;
}

/**
 * Full lead detail with all fields.
 */
export interface LeadDetail extends PipelineLead {
  linkedin_url: string | null;
  website_url: string | null;
  phone: string | null;
  country: string | null;
  industry: string | null;
  company_size: string | null;
  enrichment_data: Record<string, unknown> | null;
  outreach_data: Record<string, unknown> | null;
  tags: string[];
  notes: string | null;
  assigned_to: string | null;
  next_followup_at: string | null;
  converted_at: string | null;
  lost_reason: string | null;
  updated_at: string;
  activities: LeadHistoryEntry[];
}

/**
 * Pipeline summary for dashboard header.
 */
export interface PipelineSummary {
  total_leads: number;
  status_counts: Record<string, number>;
  conversion_rate: number;
  avg_days_in_pipeline: number;
  followup_count: number;
}

/**
 * Paginated lead list response.
 */
export interface PipelineLeadListResponse {
  leads: PipelineLead[];
  total_count: number;
  has_more: boolean;
  next_cursor: string | null;
}

/**
 * Lead activity history entry.
 */
export interface LeadHistoryEntry {
  id: string;
  lead_id: string;
  activity_type: string;
  description: string;
  timestamp: string;
  actor: string;
  metadata: Record<string, unknown> | null;
}

/**
 * Lead history response.
 */
export interface LeadHistoryResponse {
  entries: LeadHistoryEntry[];
}

/**
 * Status update request.
 */
export interface StatusUpdateRequest {
  new_status: string;
  reason?: string | null;
  note?: string | null;
}

/**
 * Status update response.
 */
export interface StatusUpdateResponse {
  success: boolean;
  lead: PipelineLead;
  message: string;
}

/**
 * Pipeline conversion metrics.
 */
export interface PipelineMetrics {
  conversion_rate: number;
  total_leads: number;
  converted_count: number;
  avg_time_per_stage: Record<string, number>;
  weekly_trend: WeeklyTrend[];
}

/**
 * Weekly trend data point.
 */
export interface WeeklyTrend {
  week: string;
  new_leads: number;
  converted: number;
}

/**
 * Follow-up candidate (from /followups endpoint).
 */
export interface FollowUpCandidate extends PipelineLead {
  needs_followup: true;
  days_since_contact: number;
}

/**
 * Add note request.
 */
export interface AddNoteRequest {
  note: string;
}

/**
 * Valid manual transitions mapping.
 * Maps current status to allowed next statuses.
 */
export const VALID_TRANSITIONS: Record<string, string[]> = {
  [LeadStatus.NEW]: [LeadStatus.RESEARCHING, LeadStatus.LOST],
  [LeadStatus.RESEARCHING]: [LeadStatus.QUALIFIED, LeadStatus.LOST],
  [LeadStatus.QUALIFIED]: [LeadStatus.OUTREACH_PENDING, LeadStatus.LOST],
  [LeadStatus.OUTREACH_PENDING]: [LeadStatus.CONTACTED, LeadStatus.LOST],
  [LeadStatus.CONTACTED]: [LeadStatus.REPLIED, LeadStatus.LOST, LeadStatus.NURTURE],
  [LeadStatus.REPLIED]: [LeadStatus.MEETING_SCHEDULED, LeadStatus.LOST, LeadStatus.NURTURE],
  [LeadStatus.MEETING_SCHEDULED]: [LeadStatus.NEGOTIATING, LeadStatus.LOST, LeadStatus.NURTURE],
  [LeadStatus.NEGOTIATING]: [LeadStatus.CONVERTED, LeadStatus.LOST, LeadStatus.NURTURE],
  [LeadStatus.CONVERTED]: [],
  [LeadStatus.LOST]: [LeadStatus.NURTURE],
  [LeadStatus.NURTURE]: [LeadStatus.CONTACTED],
};
